### ISSUE TITLE
Invisible player sprites - memory leak?

### ISSUE THREAD

**Issue by Trickiy** [TheXTech/TheXTech#779]
### Describe the issue

When I try and play any episode sometimes player sprites will not be visible even in gameplay.
![image](https://github.com/user-attachments/assets/6ae5ec64-7eca-4588-9395-9353ccfc4149)

### A type of bug

Native bug

### Does this issue happens always or randomly?

Happens randomly, rare

### Version

1.3.7-dev

### Version Hash

#bb07af2

### Platform

3DS

### Processor architecture

None

**Comment by ds-sloth** [comment]
Thanks a lot for the report!

There are two possible explanations:
(1) The game is running out of texture memory -- this is unfortunate and we can keep pushing the game's texture memory capacity up (see 32e3bb3d3d4050a0e60794e9626a0d8f05eeb021 recently pushing it from 24MB to 28MB), but this will boost the risk of instability in the rest of the game engine. I don't think that I would want to boost the texture memory past 32MB. The 3DS has more RAM than Wii but it certainly does not have unlimited RAM.
(2) Something is going wrong in the game's code intended to check for custom character sprites in the episode. Does this episode include custom character sprites?

Please share your logs so I can understand this issue further.

**Comment by ds-sloth** [comment]
Thank you very much for sharing your logs. Yes, it looks like your char3 sprite (romfs:/graphics/peach/peach-1.t3x) failed to load in the first log you shared. This is very helpful to know.

I can certainly bump the texture memory up from 28MB to 30MB, which should help with such cases. I want to be careful when doing this, because the result of running out of texture memory is that a texture fails to load, but the worst-case result of running out of main memory is a full application crash. We usually get 64MB total, and sound-related data usually take up around around 10MB, so there's not a lot of RAM left over for the actual game objects.

**Comment by Trickiy** [comment]
Could you do something like compressed cached chunks.

**Comment by ds-sloth** [comment]
Yes, our textures are currently compressed, and if you saw a case where a texture failed to load due to memory constraints, then all textures not onscreen were already evicted from cache to save memory. This is simply a hard problem: recall that SMBX 1.3 uses over 200MB of RAM all the time, and on 3DS we are managing with 64MB (and on Wii with 40MB, etc).

**Comment by Trickiy** [comment]
I watched a YouTube video a while back where a guy makes a game within 40 kb. The sprite optimizations he dose In it might reduce memory usage on 3ds if possible. Here's the [video](https://m.youtube.com/watch?v=ZWQ0591PAxM) if you wanna watch

**Comment by Wohlstand** [comment]
> I watched a YouTube video a while back where a guy makes a game within 40 kb. The sprite optimizations he dose In it might reduce memory usage on 3ds if possible. Here's the [video](https://m.youtube.com/watch?v=ZWQ0591PAxM) if you wanna watch.

Actually very interesting thing

**Comment by Wohlstand** [comment]
And, sounds like, it's possible to make a kind of the automator that will split picture to parts and will find these are equal in various orientations, and then, have in the memory the "compressed" picture, and "rules" that will be used to combine a result on the frame buffer.

**Comment by Trickiy** [comment]
This is much more severe than we initially thought because when I play for a extended period of time like an hour. The sprites start to disapear this bug being time based makes me think it's a memory leak.

**Comment by Trickiy** [comment]
here's my debug log 
[TheXTech_log_2024_08_24_08_45_48.txt](https://github.com/user-attachments/files/16736941/TheXTech_log_2024_08_24_08_45_48.txt)
towards the bottom the memory runs out.

**Comment by ds-sloth** [comment]
It's possible that this is a "genuine" memory leak, where memory is allocated and then forgotten, but there are three other possibilities -- (1) there could simply be lots of textures onscreen at once at the time that the texture load failures occur, and you hit a section with that many textures near the end of your play session, (2) load failures are permanent, so it's possible that they are simply accumulating over time, or (3) this could be memory fragmentation of the linear heap.

If this is a genuine memory leak, then it should not be fixed by changing asset packs at the main menu. If it is not a genuine memory leak, then changing asset packs will unload *everything*, and reset the load failure markers.

I'll add some additional tracking code that will help us check these cases further. After that commit arrives, please share another log if you find a similar situation in the future.

**Comment by ds-sloth** [comment]
I've sent the commit now. Another long from extended play will be very helpful.

**Comment by Trickiy** [comment]
When I am in the mood I will continue my The Invasion 2 run, if Mario's sprites start dissapearing again I will report.

**Comment by ds-sloth** [comment]
Thank you!

**Comment by Trickiy** [comment]
Ok I did some tests, test 1 being me leaving the game running on the title screen for a few hours nothing changed. This tells that there's no memory leak Wich is good. 

Test 2 continuing my The Invasion 2 run, I played for longer than an hour so its not time based.
[TheXTech_log_2024_08_24_18_10_35.txt](https://github.com/user-attachments/files/16740676/TheXTech_log_2024_08_24_18_10_35.txt) 

Test 3 restart my The Invasion 2 run, after restarting the sprites started to dissapear after I was about 38% complete. Now we know that there is leftover code in memory in every load, restart, and maybe death aswell.
[TheXTech_log_2024_08_25_09_04_00.txt](https://github.com/user-attachments/files/16740786/TheXTech_log_2024_08_25_09_04_00.txt)

**Comment by ds-sloth** [comment]
Thanks very much for updating your game and sharing your logs. When I looked at this log, I only found load failures at the title screen. I also looked again at your older log, and it also seems to only have load failures at the title screen. The load failures then carry over to levels due to the game marking them as permanent.

This is good news. I just added some code that will cause the game to aggressively try to reload these failed sprites. If texture memory becomes more free (as it will when you enter a level from the menu), then the player sprites should get reloaded within around 5 seconds.
